Overhaul of the Job entity code:

krishmaj opened the pull request:
- made the naming consistent ... workflows were always the same thing as jobs, and now all under the latter
- cleaned up the backend library and routes code for jobs (partially addressing issue #10) - removed unused functions and code, console and logger statements, and added better comments
- cleaned up and improved layouts on frontend pug pages, and renamed files for consistency
- moved Job History into the main Job Information viewing page, since there is generally so little in the former that its not worth having a separate page for it
- removed the ability to make draft jobs, in favour of allowing jobs to be edited (like components) ... for right now, this seems more appropriate, but can re-address in the future if needed
- renamed the Test Information viewing pug file to be consistent with the equivalent Component and Job ones

krishmaj committed:
Overhaul of the Job entity code:
- made the naming consistent ... workflows were always the same thing as jobs, and now all under the latter
- cleaned up the backend library and routes code for jobs (partially addressing issue #10) - removed unused functions and code, console and logger statements, and added better comments
- cleaned up and improved layouts on frontend pug pages, and renamed files for consistency
- moved Job History into the main Job Information viewing page, since there is generally so little in the former that its not worth having a separate page for it
- removed the ability to make draft jobs, in favour of allowing jobs to be edited (like components) ... for right now, this seems more appropriate, but can re-address in the future if needed
- renamed the Test Information viewing pug file to be consistent with the equivalent Component and Job ones

krishmaj commented:
@micchickenburger Verification is still not working FYI - as before, I wasn't asked for any passphrase on my end while creating the commit.

krishmaj commented:
Forgot to mention in my commit message - there are 2 functions in the `Jobs.js` file: `search` and `save` which I have not cleaned up.  These are used exclusively by the API side, which I have absolutely no idea about yet, so I don't want to make changes without a way to test things out afterwards.

krishmaj committed:
Additional changes to improve buttons and colours:
- made the buttons for creating, editing, etc. more consistent in colour and type across all job pages
- slight change to the border around the editing form to make it stand out more
- slight changes to the colours on the home page to make them consistent
- removed the diagonal colour bar across all pages ... think it looked a bit offputting on certain pages

krishmaj committed:
Removing silent errors in job routes ... should now display some sort of error message instead of just moving on.

krishmaj committed:
Fixing a couple of small typos in comments.

krishmaj committed:
Some more very small formatting and layout fixes ... changed some column widths on job list pages, removed some extraneous empty lines in code and added a bit of whitespace at the bottom of a list.

micchickenburger commented:
> @micchickenburger Verification is still not working FYI - as before, I wasn't asked for any passphrase on my end while creating the commit.

I would go over the steps again in my blog post:  https://www.micah.soy/posts/setting-up-git-identities/.  In particular, the global config setting `user.useConfigOnly` should be set to `true`.  This will stop any commits without a configured identity.  Also make sure any global identity is unset per step one.

micchickenburger commented on pug/edit_jobTypeForm.pug:
```diff
@@ -0,0 +1,198 @@
+    
+    var builder_wizard = false;
+
+    var metaschema = {"components": [{"label": "Columns",
```
This indentation feels inconsistent and hard to read.  Typical formatting in Javascript is like so:
```javascript
let someArray = [{
  key: [{
    foo: 'bar',
    baz: 'foo',
  }, {
    foo: 'bar',
    baz: 'foo',
  }],
}];
```

krishmaj replied:
I think that's basically how it was originally actually, but I found that way very hard to read myself ... personally can't stand opening brackets at one indent level on a non-empty line, and closing them at a different level on an empty one (just a pet peeve really, but a strong one!).  I'll play around and perhaps find a neater looking compromise.

krishmaj replied:
Made it more Javascript-like now.

micchickenburger commented on pug/job.pug:
```diff
@@ -0,0 +1,98 @@
+    
+    // Get the job's type form, and populate it with the job's actual submission
+    // Then disable the submission button (since the form is only to be displayed, not used)
+    window.onload = async function()
```
This will break any other window onload function that might exist.  Use `window.addEventListener('load', function);` instead.

krishmaj replied:
So just to be clear, something like this?

```
window.addEventListener('load', populateForm);

async function populateForm()
{
  ...
};
```

micchickenburger replied:
Yup!

krishmaj replied:
Done.

micchickenburger commented on pug/job.pug:
```diff
@@ -0,0 +1,98 @@
+    
+    // Get information about the job and its corresponding type form
+    var job = !{JSON.stringify(job || {}, null, 4)};
+    var form = !{JSON.stringify(form || {}, null, 4)};
```
I don’t recognize this syntax… it looks like casting from string to boolean.  Is this a pug interpolation feature of some kind?  Maybe a comment here would be useful to describe what’s happening.

krishmaj replied:
Yes, `!{}` is pug string interpolation.  I'll add comments at all instances of it to describe.

micchickenburger replied:
Okay, if it’s a normal pug thing then maybe comments won’t be necessary.  :)

micchickenburger commented on pug/job_submit.pug:
```diff
@@ -0,0 +1,138 @@
+    
+    // Get information about the job and corresponding type form
+    var job = !{JSON.stringify(job || {}, null, 4)}; 
+    var form = !{JSON.stringify(form || {}, null, 4)}; 
```
Same as above…  What is `!{}` supposed to do?

krishmaj replied:
As above.

micchickenburger commented on pug/job_submit.pug:
```diff
@@ -0,0 +1,138 @@
+        function postFail(res, statusCode, statusMsg)
+        {
+          // Set up and relay error messages
+          if(res.responseText && res.responseText.length > 0)
```
These conditions are redundant.  Empty string is falsey in JavaScript.

krishmaj replied:
Removed second condition, so it's now just `if(res.responseText)`.

micchickenburger commented on pug/job_submit.pug:
```diff
@@ -0,0 +1,138 @@
+            jobForm.setAlert("danger", res.responseText);
+          }
+          else
+          {
```
String interpolation could be used here to improve readability.

krishmaj replied:
For the `jobForm.setAlert("danger", statusMsg + " (" + statusCode + ")");` line you mean?

micchickenburger commented on pug/list_jobTypes.pug:
```diff
@@ -0,0 +1,123 @@
+        
+        $('input.NewJobTypeName').on("keyup",function(e)
+        {
+          if(e.key == "Enter")
```
Should always use strict compare `===` to avoid automatic type casting

krishmaj replied:
Changed.

micchickenburger commented on pug/job_submit.pug:
```diff
@@ -0,0 +1,138 @@
+      }
+      catch(e)
+      {
+        // TOFIX: not sure what to put here (used to be console and debugger, but we're no longer using them)
```
If we don’t have a logging solution of any kind maybe we should remove the try…catch statement altogether.  Removing the statement will allow errors to throw, which will be visible in the console but may also break something in the UI.  Leaving the catch in hides the errors.

micchickenburger commented on routes/draftRoutes.js:
```diff
@@ -14,109 +10,112 @@ var router = express.Router();
   }
   catch(err)
   {
-    logger.error(err);
```
Sorry, I should have clarified!  Server-side errors write to `stderr`!  It’s browser/client-side console statements that should be removed.  If we need to log into a server to troubleshoot an issue, logging information is helpful.

krishmaj replied:
So then I shouldn't get rid of the `logger` statements?  Only the `console` and `debugger` ones?

micchickenburger replied:
Yeah :)

krishmaj replied:
I've added the logger error messages back in where appropriate.  Note that I've done so only in `routes/jobRoutes.js` ... I've actually already removed this specific file (`routes/draftRoutes.js`) as part of the next commit I'm going to make.

micchickenburger commented on routes/jobRoutes.js:
```diff
@@ -1,155 +1,220 @@
+    // Throw an error if there is no DB entry corresponding to this job ID
+    var job = await Jobs.retrieve(req.params.jobId);
+    
+    if(!job)
```
`if (!job) next();` is more succinct

krishmaj replied:
Mmm, personally I still prefer even single-line statements to be indented in brackets, same as multi-line blocks.  Feel like it keeps things more consistent.  Chalk it up to stylistic differences?

micchickenburger commented on static/css/dashboard.css:
```diff
@@ -211,28 +211,16 @@ body.deployment-devsite .navbar-brand img {
 body.deployment-laptop #navbar-brand
 {
-  background-color: #f7f7d7 !important;
+  background-color: #fff3cd !important;
```
Oh boy.  `!important` declarations in CSS are a major headache and slow page render performance.  We will have to tackle these at some point.

krishmaj replied:
Sure.  They were already present, so I just left them as they are.  I'll link this comment to the CSS related issue as well.

micchickenburger reviewed (dismissed):
Overall all looks fine.  Just nitpicks.  There is a lot of C-style here, which is atypical in the JavaScript/webdev world, but we can auto-address later when we get a linter in place.  Lots of `pre` statements, but you already know how I feel about those.  😁

krishmaj committed:
Changes from PR review:
- changed indentation formatting to be more Javascript-like in edit_jobTypeForm.pug
- changed onload function to be addEventListener + separate function in job.pug
- removed redundant res.responseText condition from job_submit.pug
- changed to using strict comparison (===) in list_jobTypes.pug
- put back the logger messages for errors only in jobRoutes (logger info messages are still removed)

krishmaj commented:
@micchickenburger I tried doing the Github identities stuff again yesterday, but it apparently hasn't made any difference.  I've sent you an email about that, but it looks like there's issues with my email service right now ... delays on sending and receiving.  (I only just got notifications a few minutes ago about your comments on here from yesterday!).  So who knows when you'll get it.
